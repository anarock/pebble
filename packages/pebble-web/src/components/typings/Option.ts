import { ControlProps } from "./Control";
import { Omit } from "utility-types";
import * as React from "react";
import { Interpolation } from "./emotionCustom";

export interface OptionProps<OptionType>
  extends Omit<ControlProps<OptionType>, "checked" | "type"> {
  isActive?: boolean;
  isSelected?: boolean;
  multiSelect?: boolean;
  labelStyles?: Interpolation;
  rightElement?: (props: OptionProps<OptionType>) => React.ReactNode;
}
