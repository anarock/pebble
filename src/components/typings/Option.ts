import { ControlProps } from "./Control";
import { Omit } from "utility-types";
import * as React from "react";

export interface OptionProps<OptionType>
  extends Omit<ControlProps<OptionType>, "checked" | "type"> {
  isActive?: boolean;
  isSelected?: boolean;
  multiSelect?: boolean;
  labelClassName?: string;
  rightElement?: (props: OptionProps<OptionType>) => React.ReactNode;
}
