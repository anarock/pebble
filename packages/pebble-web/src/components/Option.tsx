import * as React from "react";
import { OptionProps } from "./typings/Option";
import Control from "./shared/Control";
import {
  activeRow,
  rowWrapper,
  selectedRow,
  labelWrap
} from "./styles/Options.styles";
import Ink from "react-ink";
import { colors } from "pebble-shared";

const defaultProps = {
  rightElement: ({
    isSelected,
    multiSelect
  }: {
    isSelected: boolean;
    multiSelect: boolean;
  }) => {
    const iconClass = [
      "pi",
      isSelected && "pi-checkbox-selected",
      !isSelected && "pi-checkbox-unselected"
    ]
      .filter(Boolean)
      .join(" ");
    return multiSelect ? (
      <i
        className={iconClass}
        css={{
          marginLeft: "10px",
          color: isSelected ? colors.violet.base : colors.gray.light,
          fontSize: "20px"
        }}
      />
    ) : null;
  }
};

class Option<OptionType> extends React.Component<
  OptionProps<OptionType> & Required<typeof defaultProps>
> {
  static defaultProps = defaultProps;
  render() {
    const {
      label,
      isActive,
      isSelected,
      rightElement,
      labelStyles,
      styles
    } = this.props;
    const _class = [
      rowWrapper,
      !!isActive && activeRow,
      !!isSelected && selectedRow,
      styles
    ];
    return (
      <Control
        {...this.props}
        checked={this.props.isSelected}
        type={this.props.multiSelect ? "checkbox" : "radio"}
        styles={_class}
      >
        {() => {
          return (
            <>
              <div css={[labelWrap, labelStyles]}>{label}</div>
              {rightElement(this.props)}
              <Ink />
            </>
          );
        }}
      </Control>
    );
  }
}

export default Option;
