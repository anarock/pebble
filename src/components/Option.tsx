import * as React from "react";
import { OptionProps } from "./typings/Option";
import Control from "./shared/Control";
import { cx } from "emotion";
import { activeRow, rowWrapper, selectedRow } from "./styles/Options.styles";
import Ink from "react-ink";
import { colors, mixins } from "../theme";
import { Component } from "react";

class Option extends Component<OptionProps> {
  static defaultProps: Partial<OptionProps> = {
    rightElement: ({ isSelected, multiSelect }) => {
      const iconClass = cx({
        "icon-checkbox-selected": isSelected,
        "icon-checkbox-unselected": !isSelected
      });
      return multiSelect ? (
        <i
          style={{
            color: isSelected ? colors.violet.base : colors.gray.light
          }}
          className={iconClass}
        />
      ) : null;
    }
  };

  render() {
    return (
      <Control
        {...this.props}
        checked={this.props.isSelected}
        type={this.props.multiSelect ? "checkbox" : "radio"}
      >
        {() => {
          const { label, isActive, isSelected, rightElement } = this.props;

          const _class = cx(rowWrapper, {
            [activeRow]: isActive,
            [selectedRow]: isSelected
          });

          return (
            <div className={_class} style={mixins.flexSpaceBetween}>
              {label} {rightElement(this.props)}
              <Ink />
            </div>
          );
        }}
      </Control>
    );
  }
}

export default Option;
