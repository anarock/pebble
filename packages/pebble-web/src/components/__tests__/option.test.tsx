import * as React from "react";
import Option from "../Option";
import renderer from "react-test-renderer";

describe("Option component", () => {
  test("snapshot in selected state", () => {
    const checkbox = renderer.create(
      <Option
        isSelected
        onChange={() => {}}
        value="option"
        label="I am an option"
      />
    );

    const tree = checkbox.toJSON();
    expect(tree).toMatchSnapshot();
  });

  test("snapshot in active state", () => {
    const checkbox = renderer.create(
      <Option
        isActive
        onChange={() => {}}
        value="option"
        label="I am an option"
      />
    );

    const tree = checkbox.toJSON();
    expect(tree).toMatchSnapshot();
  });

  test("snapshot for multi select", () => {
    const checkbox = renderer.create(
      <Option
        multiSelect
        onChange={() => {}}
        value="option"
        label="I am an option"
      />
    );

    const tree = checkbox.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
