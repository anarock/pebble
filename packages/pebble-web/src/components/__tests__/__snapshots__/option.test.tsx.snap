// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Option component snapshot for multi select 1`] = `
.emotion-0 {
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  overflow: hidden;
  text-overflow: ellipsis;
}

.emotion-2 {
  cursor: pointer;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  outline: none;
  padding: 10px 0;
  position: relative;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  font-weight: 400;
  color: #101721;
  font-size: 14px;
  font-weight: 400;
  color: #101721;
  font-size: 14px;
  cursor: pointer;
  padding: 12px 18px;
  position: relative;
  line-height: 28px;
  z-index: 999;
  overflow-x: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  -webkit-align-content: initial;
  -ms-flex-line-pack: initial;
  align-content: initial;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.emotion-2[data-disabled='true'] {
  cursor: not-allowed;
}

.emotion-2:last-of-type {
  border: 0;
}

.emotion-2:hover {
  background-color: #F7F7F9;
}

.emotion-1 {
  margin-left: 10px;
  color: #E0E0E0;
  font-size: 20px;
}

<div
  className="emotion-2"
  onClick={[Function]}
  role="checkbox"
  tabIndex={-1}
>
  <div
    className="emotion-0"
  >
    I am an option
  </div>
  <i
    className="pi pi-checkbox-unselected emotion-1"
  />
  <canvas
    height={0}
    onDragOver={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    style={
      Object {
        "borderRadius": "inherit",
        "height": "100%",
        "left": 0,
        "position": "absolute",
        "top": 0,
        "width": "100%",
      }
    }
    width={0}
  />
</div>
`;

exports[`Option component snapshot in active state 1`] = `
.emotion-0 {
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  overflow: hidden;
  text-overflow: ellipsis;
}

.emotion-1 {
  cursor: pointer;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  outline: none;
  padding: 10px 0;
  position: relative;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  font-weight: 400;
  color: #101721;
  font-size: 14px;
  font-weight: 400;
  color: #101721;
  font-size: 14px;
  cursor: pointer;
  padding: 12px 18px;
  position: relative;
  line-height: 28px;
  z-index: 999;
  overflow-x: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  -webkit-align-content: initial;
  -ms-flex-line-pack: initial;
  align-content: initial;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background-color: #F7F7F9;
}

.emotion-1[data-disabled='true'] {
  cursor: not-allowed;
}

.emotion-1:last-of-type {
  border: 0;
}

.emotion-1:hover {
  background-color: #F7F7F9;
}

<div
  className="emotion-1"
  onClick={[Function]}
  role="radio"
  tabIndex={-1}
>
  <div
    className="emotion-0"
  >
    I am an option
  </div>
  <canvas
    height={0}
    onDragOver={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    style={
      Object {
        "borderRadius": "inherit",
        "height": "100%",
        "left": 0,
        "position": "absolute",
        "top": 0,
        "width": "100%",
      }
    }
    width={0}
  />
</div>
`;

exports[`Option component snapshot in selected state 1`] = `
.emotion-1 {
  cursor: pointer;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  outline: none;
  padding: 10px 0;
  position: relative;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  font-weight: 400;
  color: #101721;
  font-size: 14px;
  font-weight: 400;
  color: #101721;
  font-size: 14px;
  cursor: pointer;
  padding: 12px 18px;
  position: relative;
  line-height: 28px;
  z-index: 999;
  overflow-x: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  -webkit-align-content: initial;
  -ms-flex-line-pack: initial;
  align-content: initial;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  color: #101721;
  font-weight: 500;
  font-size: 14px;
}

.emotion-1[data-disabled='true'] {
  cursor: not-allowed;
}

.emotion-1:last-of-type {
  border: 0;
}

.emotion-1:hover {
  background-color: #F7F7F9;
}

.emotion-0 {
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  overflow: hidden;
  text-overflow: ellipsis;
}

<div
  aria-checked={true}
  className="emotion-1"
  onClick={[Function]}
  role="radio"
  tabIndex={0}
>
  <div
    className="emotion-0"
  >
    I am an option
  </div>
  <canvas
    height={0}
    onDragOver={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    style={
      Object {
        "borderRadius": "inherit",
        "height": "100%",
        "left": 0,
        "position": "absolute",
        "top": 0,
        "width": "100%",
      }
    }
    width={0}
  />
</div>
`;
